sys/time_units.h: Work around clang div-by-zero warning

keith-packard opened the pull request:
clang emits a warning for this code:

  foo.c:1:25: warning: division by zero is undefined [-Wdivision-by-zero]
  int i = (10 > 100 ? (20 / (10 / 100)) : (20 * (100 / 10)));
	                        ^ ~~~~~~~~~~

The warning is generated for code whose value does not affect the expression result, but technically it is still 'undefined behavior'.

Work around this warning by checking for a zero divisor and substituting one to make the compiler happy.

Closes: #63564

keith-packard committed:
sys/time_units.h: Work around clang div-by-zero warning

clang emits a warning for this code:

  foo.c:1:25: warning: division by zero is undefined [-Wdivision-by-zero]
  int i = (10 > 100 ? (20 / (10 / 100)) : (20 * (100 / 10)));
	                        ^ ~~~~~~~~~~

The warning is generated for code whose value does not affect the
expression result, but technically it is still 'undefined behavior'.

Work around this warning by checking for a zero divisor and substituting
one to make the compiler happy.

Signed-off-by: Keith Packard <keithp@keithp.com>

fabiobaltieri reviewed (dismissed):
Thanks!

Just a few compliance warnings:

```
 SPACING: space required after that ',' (ctx:VxV)
File:include/zephyr/sys/time_units.h
Line:146
 SPACING: spaces required around that '?:' (ctx:WxV)
File:include/zephyr/sys/time_units.h
Line:146
 SPACING: space required after that ',' (ctx:VxV)
File:include/zephyr/sys/time_units.h
Line:148
```

evgeniy-paltsev reviewed (changes requested):
Can we add a test-case which triggers this issue to `tests/kernel/timer/timer_api/src/timer_convert.c`?

keith-packard committed:
sys/time_units.h: Make sys_clock_hw_cycles_per_sec() constant if possible

To be used in a global initializer, this must be a macro instead of a
static inline when the return value is constant.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/timer_api: Make sure constant time conversions are constants

When the timer frequency is known at compile time, make sure we can use any
time conversion macro as a global initializer.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> Can we add a test-case which triggers this issue to `tests/kernel/timer/timer_api/src/timer_convert.c`?

good idea. That also caught a problem with the definition of Z_HZ_cyc, which used sys_clock_hw_cycles_per_sec. That was never a constant expression because it was a static inline instead of being a macro.

evgeniy-paltsev commented:
> That also caught a problem with the definition of Z_HZ_cyc, which used sys_clock_hw_cycles_per_sec. That was never a constant expression because it was a static inline instead of being a macro.

Hmm, do we require that sys_clock_hw_cycles_per_sec must be compile-time constant?

Cause it naturally can't be compile-time constant if `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` is enabled - so we use `sys_clock_hw_cycles_per_sec_runtime_get()`

keith-packard commented:
> > That also caught a problem with the definition of Z_HZ_cyc, which used sys_clock_hw_cycles_per_sec. That was never a constant expression because it was a static inline instead of being a macro.
> 
> Hmm, do we require that sys_clock_hw_cycles_per_sec must be compile-time constant?

No, it's only constant when it isn't read at runtime. The issue was that `sys_clock_hw_cycles_per_sec()` wasn't a "constant" even when the timer frequency was specified at compile time.

> Cause it naturally can't be compile-time constant if `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` is enabled - so we use `sys_clock_hw_cycles_per_sec_runtime_get()`

Yup. In that case, we aren't testing to make sure the timer macros are valid C constant expressions.

fgrandel reviewed (dismissed):
Nitpick: I personally try to avoid `#ifndef ... #else ...` as the negation is easy to overlook and requires extra cognitive load.

Why not `#ifdef ... #else ...` instead?

keith-packard commented:
> > That also caught a problem with the definition of Z_HZ_cyc, which used sys_clock_hw_cycles_per_sec. That was never a constant expression because it was a static inline instead of being a macro.
> 
> Hmm, do we require that sys_clock_hw_cycles_per_sec must be compile-time constant?
> 
> Cause it naturally can't be compile-time constant if `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` is enabled - so we use `sys_clock_hw_cycles_per_sec_runtime_get()`

I'm wondering if this means we need something like #63594 ?

fabiobaltieri commented:
@evgeniy-paltsev can you revisit this once you have a chance? Would be nice to merge it before RC2.

jhedberg commented:
> @evgeniy-paltsev can you revisit this once you have a chance? Would be nice to merge it before RC2.

@andyross (as the assignee) would also need to approve first

andyross commented on include/zephyr/sys/time_units.h:
```diff
@@ -138,6 +135,16 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
+ * around this by substituting 1 to make the compiler happy.
+ */
+#ifdef __clang__
+#define z_tmcvt_divisor(a, b) ((a) / (b) ?: 1)
```
At what level of the language standard did that ternary syntax appear?  I worry because we have `__clang__` compilers in the stable (c.f. Cadence's xt-clang) which are comparatively old.  Might be safer to write that out long hand?  Or not, if we're sure this is supported.

andyross replied:
Also grade school math pedantry: the operation of this macro should probably be named "quotient", or just "divide", the "divisor" is the `a` argument.

keith-packard replied:
Yeah, the name actually refers to the division that this value is used as a part of -- which is in the `z_tmcvt_int_div_*` macros. In that context, it looks like I guessed right in using 'divisor' (I admit that I just checked :-)

andyross reviewed (approved):
Some nitpicks, but not worth a -1, looks fine on the whole.

andyross commented:
> As far as I can tell, this only impacts qemu_x86 platforms (as those appear to use HPET which implies run-time timer freq).

FWIW: I wouldn't cry if someone went in there with a hatchet and just removed the runtime frequency query from the HPET driver.  Literally everything else in Zephyr is based on compile-time hardware configuration, and in point of fact all of our HPET x86 devices have hard frequency settings anyway, IIRC.

CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME is a historical wart, and it doesn't even do the thing we actually know we need, which is dynamic runtime frequency selection (right now there are a handful of devices that do that, but by translation inside the timer driver).

keith-packard commented:
> CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME is a historical wart, and it doesn't even do the thing we actually know we need, which is dynamic runtime frequency selection (right now there are a handful of devices that do that, but by translation inside the timer driver).

I'd suspect to make this work reliably you'd need to annotate every timeout value with some kind of 'required precision' value so that the dynamic frequency stuff didn't introduce so much slop that things started breaking when the core timer frequency got set too low...

In any case, it looks like HPET doesn't require runtime timer values, it just allows them, so turning all of that off could be done by simply removing the 'imply y' value from the relevant Kconfig file.
